tests/PHPStan/Rules/Variables/data/bug-4173.php

```diff
@@ -0,0 +1,34 @@
+<?php declare(strict_types = 1);
+
+namespace Bug4173;
+
+class HelloWorld
+{
+	protected function _sub_render_condition($row)
+	{
+		if (count($row) === 3) {
+			[$field, $cond, $value] = $row;
+		} elseif (count($row) === 2) {
+			[$field, $cond] = $row;
+		} elseif (count($row) === 1) {
+			[$field] = $row;
+		} else {
+			throw new \InvalidArgumentException();
+		}
+
+		$field = trim($field);
+
+		if (count($row) === 1) {
+			return $field;
+		}
+
+		if (count($row) === 2) {
```

Comment:
This PR seems to cache result of function result 1:1 and retrive the result from cache later. I observed this in phpstan v 1.9.2 (the latest rel).

Is that desired? For `count()` it is legit and thank you. However for expression like `TraitUtil::hasTrait($object, DebugTrait::class)` (a custom function) how does php know/guarantee the result will be the same, does it reason it based on the `TraitUtil::hasTrait` impl or just guess it when the passed args are the same?

Reply:
Yes this is intended

> how does php know/guarantee the result will be the same

When
* TraitUtil::hasTrait is not marked impure
* The exact same `TraitUtil::hasTrait($object, DebugTrait::class)` is called
* $object is not invalidated

> just guess it when the passed args are the same?

so it's something like that.

Reply:
By the way, remembering function results was working before this PR already.
https://phpstan.org/blog/remembering-and-forgetting-returned-values
this PR made them work for more complex conditions too.

Reply:
Thank you, I did not know user functions are "pure"/remembered by default.

Reply:
See https://phpstan.org/blog/remembering-and-forgetting-returned-values